asf_heat_pump_suitability/analysis/exploratory/create_full_dataset/assign_cluster_suitability_and_feasibility.py

```diff
@@ -2,236 +2,402 @@
-    hn_area: bool,
-    city_centre: bool,
-    has_outdoor_space: bool,
-) -> str:
+
+def prepare_df_for_suitability_categorisation(
+    df: pl.DataFrame,
+    city_centre_oas: set = config.city_centre_oas,
+    outdoor_space_threshold: float = config.outdoor_space_threshold,
+) -> pl.DataFrame:
+    """
+    Prepares the dataframe for suitability categorisation. This includes:
+    - Identifying clusters close to the city center or in heat network zones
+    - Identifying clusters with outdoor space
+
+    Args:
+        df (pl.DataFrame): DataFrame containing the dataset with feature values for each UPRN.
+        city_centre_oas (set, optional): OAs that are considered part of the city center. Should be an oa21 format.
+            Defaults to config.city_centre_oas.
+        outdoor_space_threshold (float, optional): Threshold for outdoor space in meters squared.
+        i.e. if garden_area_m2 > 0, then has_outdoor_space = True
+
+    Returns:
+        pl.DataFrame: DataFrame ready for suitability categorisation.
+    """
+    # Create city_centre column based on whether the UPRN is in the set of city centre OAs
+    df = df.with_columns(pl.col("oa21").is_in(city_centre_oas).alias("in_city_centre"))
+
+    # Transform predicted_tenure and predicted_property_type into dummies
+    df = df.to_dummies("predicted_property_type")
+    df = df.rename({"predicted_property_type_Flat, maisonette or apartment": "flats"})
+
+    cluster_df = df.group_by("cluster").agg(
+        # Cluster size
+        pl.col("UPRN").count().alias("cluster_size"),
+        # Heat network zone logic: if at least one property in the cluster is in a heat network zone,
+        # the whole cluster is considered to be in a heat network zone
+        pl.col("in_heat_network_zone").max().alias("in_heat_network_zone"),
+        # City centre logic: if at least one property in the cluster is in the city centre,
+        # the whole cluster is considered to be close to the city centre
+        pl.col("in_city_centre").max().alias("in_city_centre"),
+        # Outdoor space logic:
+        (
+            # Properties other than flats: sum garden_area_m2 for all properties
+            pl.col("garden_area_m2")
+            .filter(pl.col("flats") != 1)
+            .sum()
+            .fill_null(0)
+        ).alias("total_outdoor_space"),
+    )
+
+    # Calculate garden sizes for flats
+    flats_garden_sizes = (
+        df.group_by("cluster")
+        .agg(
+            (
+                (
+                    pl.struct(["building_id", "garden_area_m2"])
+                    .filter(pl.col("flats") == 1)
+                    .gather(
+                        pl.col("building_id").filter(pl.col("flats") == 1).arg_unique()
+                    )
+                    .struct.field("garden_area_m2")
+                )
+            )
+        )
+        .with_columns(
+            pl.col("building_id").list.sum().alias("total_outdoor_space_flats")
+        )
+    )
```

Comment:
```suggestion
flats_garden_sizes = df.filter(pl.col("flats") == 1).group_by(["cluster", "building_id"]).agg(pl.col("garden_area_m2").unique().sum().alias("total_outdoor_space_flats"))
```
I think this does the same thing (?) in a simpler way? The code I've written is getting the total garden size for every unique building ID in a cluster.

Reply:
I think this is similar to what I had originally, but it actually outputs something different. Liz spent some time looking at this methodology, to make sure it works correctly, so I'll leave her version for now and maybe later we can spend time improving the code (but making sure the computations are correct).

Reply:
Ok no worries, in that case it would be great to add some documentation here to the lines within the `.agg` method.